--- glibc-2.3.6/timezone/tzfile.old	2005-04-06 11:20:45.000000000 +1000
+++ glibc-2.3.6/timezone/tzfile.h	2008-09-11 07:20:54.000000000 +1000
@@ -37,6 +37,10 @@
 #define TZDEFAULT	"localtime"
 #endif /* !defined TZDEFAULT */
 
+#ifndef POZIX_TZ_FILE
+#define POSIX_TZ_FILE 	"/etc/config/TZ"
+#endif /* !defined POSIX_TZ_FILE */
+
 #ifndef TZDEFRULES
 #define TZDEFRULES	"posixrules"
 #endif /* !defined TZDEFRULES */
