--- glibc-2.3.6/csu/Makefile.orig	2008-09-05 09:20:01.000000000 +1000
+++ glibc-2.3.6/csu/Makefile	2008-09-05 09:20:08.000000000 +1000
@@ -241,7 +241,7 @@
 	 esac; \
 	 files="$(all-Banner-files)";				\
 	 if test -n "$$files"; then				\
-	   echo "\"Available extensions:\\n\"";			\
+	   /bin/echo "\"Available extensions:\\n\"";			\
 	   sed -e '/^#/d' -e 's/^[[:space:]]*/	/'		\
 	       -e 's/^\(.*\)$$/\"\1\\n\"/' $$files;		\
 	 fi) > $@T
