--- glibc-2.3.6/configure.old	2008-07-12 14:02:28.000000000 +1000
+++ glibc-2.3.6/configure	2008-07-12 14:02:30.000000000 +1000
@@ -6304,7 +6304,7 @@
   ac_status=$?
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); }; } ||
-   { ac_try='$libc_unwind_check -lgcc_eh -lgcc >&5'
+   { ac_try='$libc_unwind_check -lgcc >&5'
   { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
   (eval $ac_try) 2>&5
   ac_status=$?
--- glibc-2.3.6/Makeconfig.old	2008-07-12 14:01:35.000000000 +1000
+++ glibc-2.3.6/Makeconfig	2008-07-12 14:01:37.000000000 +1000
@@ -503,12 +503,12 @@
   libunwind = -lunwind
 endif
 ifneq ($(have-as-needed),yes)
- libgcc_eh := -lgcc_eh $(libunwind)
+ libgcc_eh := $(libunwind)
 else
  libgcc_eh := -Wl,--as-needed -lgcc_s$(libgcc_s_suffix) $(libunwind) -Wl,--no-as-needed
 endif
 gnulib := -lgcc $(libgcc_eh)
-static-gnulib := -lgcc -lgcc_eh $(libunwind)
+static-gnulib := -lgcc $(libunwind)
 libc.so-gnulib := -lgcc
 endif
 ifeq ($(elf),yes)
