--- glibc-2.3.6/sysdeps/unix/sysv/linux/bits/in.h.orig	2009-08-20 16:46:29.000000000 +1000
+++ glibc-2.3.6/sysdeps/unix/sysv/linux/bits/in.h	2009-08-20 16:47:25.000000000 +1000
@@ -62,6 +62,12 @@
 #define IP_RECVERR         11	/* bool */
 #define IP_RECVTTL         12	/* bool */
 #define IP_RECVTOS         13	/* bool */
+#define IP_MTU             14
+#define IP_FREEBIND        15
+#define IP_IPSEC_POLICY    16
+#define IP_XFRM_POLICY     17
+#define IP_PASSSEC         18
+#define IP_TRANSPARENT     19
 
 
 /* IP_MTU_DISCOVER arguments.  */
