diff -uNr pcmcia-cs-3.1.34.orig/wireless/Makefile pcmcia-cs-3.1.34/wireless/Makefile
--- pcmcia-cs-3.1.34.orig/wireless/Makefile	Mon Aug 27 06:14:13 2001
+++ pcmcia-cs-3.1.34/wireless/Makefile	Sun Jul 21 16:57:30 2002
@@ -23,9 +23,13 @@
 SRCS    += hermes.c orinoco.c orinoco_cs.c
 endif
 
+# Host AP driver
+MODULES	+= hostap.o hostap_crypt_wep.o hostap_cs.o
+SRCS	+= hostap.c hostap_crypt_wep.c hostap_cs.c
+
 all:	$(MODULES)
 
-wvlan_hcf.o wvlan_hcfio.o hermes.o orinoco.o: %.o: %.c
+wvlan_hcf.o wvlan_hcfio.o hermes.o orinoco.o hostap.o: %.o: %.c
 	$(CC) -MD -c $(XFLAGS) $<
 	@mkdir -p .depfiles ; mv $*.d .depfiles
 
