diff -uar boost_1_39_0/boost/config/stdlib/stlport.hpp build/boost_1_39_0/boost/config/stdlib/stlport.hpp
--- boost_1_39_0/boost/config/stdlib/stlport.hpp	2009-04-07 07:25:18.000000000 +1000
+++ build/boost_1_39_0/boost/config/stdlib/stlport.hpp	2010-10-12 12:35:18.761958748 +1000
@@ -129,8 +129,8 @@
          || defined(_STLP_IMPORT_VENDOR_CSTD) \
          || defined(_STLP_USE_OWN_NAMESPACE)) \
       && (defined(__STL_VENDOR_GLOBAL_CSTD) || defined (_STLP_VENDOR_GLOBAL_CSTD))
-#     define BOOST_NO_STDC_NAMESPACE
-#     define BOOST_NO_EXCEPTION_STD_NAMESPACE
+//#     define BOOST_NO_STDC_NAMESPACE
+//#     define BOOST_NO_EXCEPTION_STD_NAMESPACE
 #  endif
 #elif defined(__BORLANDC__) && __BORLANDC__ < 0x560
 // STLport doesn't import std::abs correctly:
