This definition conflicts with the one in uClibc 0.9.29.

--- STLport-5.2.1.orig/stlport/stl/_cmath.h	2008-12-10 19:56:51.000000000 +1000
+++ STLport-5.2.1/stlport/stl/_cmath.h	2010-10-06 17:14:52.152864924 +1000
@@ -500,10 +500,7 @@
 #    pragma warning (pop)
 #  endif
 #else
-#  if defined (_STLP_USE_UCLIBC)
-inline double hypot(double x, double y) { return sqrt(x * x + y * y); }
-_STLP_DEF_MATH_INLINE2(hypot, hypot)
-#  elif defined (_STLP_WCE)
+#  if defined (_STLP_WCE)
    /* CE has a double _hypot(double,double) which we use */
 inline double hypot(double __x, double __y) { return _hypot(__x,__y); }
 _STLP_DEF_MATH_INLINE2(hypot, _hypot)
