diff -uar cryptopp561/GNUmakefile build/cryptopp561/GNUmakefile
--- cryptopp561/GNUmakefile	2010-08-09 14:22:42.000000000 +1000
+++ build/cryptopp561/GNUmakefile	2010-10-11 15:22:41.281936326 +1000
@@ -1,4 +1,4 @@
-CXXFLAGS = -DNDEBUG -g -O2
+#CXXFLAGS = -DNDEBUG -g -O2
 #CXXFLAGS = -g
 # -fPIC is supported. Please report any breakage of -fPIC as a bug.
 # CXXFLAGS += -fPIC
@@ -11,9 +11,10 @@
 MKDIR = mkdir
 EGREP = egrep
 UNAME = $(shell uname)
-ISX86 = $(shell uname -m | $(EGREP) -c "i.86|x86|i86|amd64")
+#ISX86 = $(shell uname -m | $(EGREP) -c "i.86|x86|i86|amd64")
 
 # Default prefix for make install
+PREFIX ?= $(DESTDIR)
 ifeq ($(PREFIX),)
 PREFIX = /usr
 endif
@@ -69,6 +70,8 @@
 LDLIBS += -lws2_32
 endif
 
+else
+IS_SUN_CC = 0
 endif	# ISX86
 
 ifeq ($(UNAME),)	# for DJGPP, where uname doesn't exist
@@ -131,7 +134,7 @@
 TESTIMPORTOBJS = $(TESTOBJS:.o=.import.o)
 DLLTESTOBJS = dlltest.dllonly.o
 
-all: cryptest.exe
+all: cryptest.exe libcryptopp.a libcryptopp.so
 
 test: cryptest.exe
 	./cryptest.exe v
