Date: 2006/11/11 02:55:24
#11908

Increase maximum configurable limit of cachesize to 50k.

Index: dnsmasq2/src/option.c
===================================================================
--- dnsmasq2.orig/src/option.c	2009-03-12 13:27:41.000000000 +1000
+++ dnsmasq2/src/option.c	2009-03-12 13:29:33.000000000 +1000
@@ -1497,8 +1497,8 @@ static char *one_opt(int option, char *a
 	    
 	    if (size < 0)
 	      size = 0;
-	    else if (size > 10000)
-	      size = 10000;
+	    else if (size > 50000)
+	      size = 50000;
 	    
 	    daemon->cachesize = size;
 	  }
