diff -uNr linux-2.5.60.orig/MAINTAINERS linux-2.5.60/MAINTAINERS
--- linux-2.5.60.orig/MAINTAINERS	2003-02-11 19:31:52.000000000 -0800
+++ linux-2.5.60/MAINTAINERS	2003-02-11 19:34:04.000000000 -0800
@@ -735,6 +735,13 @@
 L:	linux-hippi@sunsite.dk
 S:	Maintained
 
+HOST AP DRIVER
+P:	Jouni Malinen
+M:	jkmaline@cc.hut.fi
+L:	hostap@shmoo.com
+W:	http://hostap.epitest.fi/
+S:	Maintained
+
 HP100:	Driver for HP 10/100 Mbit/s Voice Grade Network Adapter Series
 P:	Jaroslav Kysela
 M:	perex@suse.cz
diff -uNr linux-2.5.60.orig/drivers/net/wireless/Kconfig linux-2.5.60/drivers/net/wireless/Kconfig
--- linux-2.5.60.orig/drivers/net/wireless/Kconfig	2003-02-11 19:31:54.000000000 -0800
+++ linux-2.5.60/drivers/net/wireless/Kconfig	2003-02-11 19:34:04.000000000 -0800
@@ -182,6 +182,87 @@
 	  for location).  You also want to check out the PCMCIA-HOWTO,
 	  available from <http://www.linuxdoc.org/docs.html#howto>.
 
+config HOSTAP
+	tristate "Host AP support for Prism2/2.5/3 IEEE 802.11b"
+	depends on NET_RADIO
+	---help---
+	A driver for 802.11b wireless cards based on Intersil Prism2/2.5/3
+	chipset. This driver supports so called Host AP mode that allows
+	the card to act as an IEEE 802.11 access point.
+
+	See <http://hostap.epitest.fi/> for more information about the
+	Host AP driver configuration and tools.
+
+	This option includes the base Host AP driver code that is shared by
+	different hardware models. You will also need to enable support for
+	PLX/PCI/CS version of the driver to actually use the driver.
+
+	The driver can be compiled as modules and they will be called
+	"hostap.o" and "hostap_crypt_wep.o".
+
+config HOSTAP_HOSTAPD
+	bool "Use user space daemon (hostapd) with Host AP driver"
+	depends on HOSTAP
+	---help---
+	Configure Host AP driver to use user space daemon (hostapd), for
+	access point management functionality. If this is not enabled, the
+	driver uses kernel side implementation of AP management.
+
+	hostapd daemon is available from http://hostap.epitest.fi/. In addition
+	to the management functionality in the kernel driver, it implements
+	IEEE 802.1x Authenticator that can be used to authenticate stations,
+	e.g., with EAP-TLS.
+
+config HOSTAP_FIRMWARE
+	bool "Support downloading firmware images with Host AP driver"
+	depends on HOSTAP
+	---help---
+	Configure Host AP driver to include support for firmware image
+	download. Current version supports only downloading to volatile, i.e.,
+	RAM memory. Flash upgrade is not yet supported.
+
+	Firmware image downloading needs user space tool, prism2_srec. It is
+	available from http://hostap.epitest.fi/.
+
+config HOSTAP_PLX
+	tristate "Host AP driver for Prism2/2.5/3 in PLX9052 PCI adaptors"
+	depends on PCI && HOSTAP
+	---help---
+	Host AP driver's version for Prism2/2.5/3 PC Cards in PLX9052 based
+	PCI adaptors.
+
+	"Host AP support for Prism2/2.5/3 IEEE 802.11b" is required for this
+	driver and its help text includes more information about the Host AP
+	driver.
+
+	The driver can be compiled as a module and will be named
+	"hostap_plx.o".
+
+config HOSTAP_PCI
+	tristate "Host AP driver for Prism2.5 PCI adaptors"
+	depends on PCI && HOSTAP
+	---help---
+	Host AP driver's version for Prism2.5 PCI adaptors.
+
+	"Host AP support for Prism2/2.5/3 IEEE 802.11b" is required for this
+	driver and its help text includes more information about the Host AP
+	driver.
+
+	The driver can be compiled as a module and will be named
+	"hostap_pci.o".
+
+config HOSTAP_CS
+	tristate "Host AP driver for Prism2/2.5/3 PC Cards"
+	depends on PCMCIA!=n && HOSTAP
+	---help---
+	Host AP driver's version for Prism2/2.5/3 PC Cards.
+
+	"Host AP support for Prism2/2.5/3 IEEE 802.11b" is required for this
+	driver and its help text includes more information about the Host AP
+	driver.
+
+	The driver can be compiled as a module and will be named "hostap_cs.o".
+
 # yes, this works even when no drivers are selected
 config NET_WIRELESS
 	bool
diff -uNr linux-2.5.60.orig/drivers/net/wireless/Makefile linux-2.5.60/drivers/net/wireless/Makefile
--- linux-2.5.60.orig/drivers/net/wireless/Makefile	2003-02-11 19:31:54.000000000 -0800
+++ linux-2.5.60/drivers/net/wireless/Makefile	2003-02-11 19:34:04.000000000 -0800
@@ -15,3 +15,8 @@
 
 obj-$(CONFIG_AIRO)		+= airo.o
 obj-$(CONFIG_AIRO_CS)		+= airo_cs.o airo.o
+
+obj-$(CONFIG_HOSTAP)		+= hostap.o hostap_crypt_wep.o
+obj-$(CONFIG_HOSTAP_CS)		+= hostap_cs.o
+obj-$(CONFIG_HOSTAP_PLX)	+= hostap_plx.o
+obj-$(CONFIG_HOSTAP_PCI)	+= hostap_pci.o
