Don't syslog about ipset rehashing, it happens too often.

Index: ipset/kernel/include/linux/netfilter_ipv4/ip_set_hashes.h
===================================================================
--- ipset.orig/kernel/include/linux/netfilter_ipv4/ip_set_hashes.h	2009-09-14 14:11:05.000000000 +1000
+++ ipset/kernel/include/linux/netfilter_ipv4/ip_set_hashes.h	2009-09-14 14:16:23.000000000 +1000
@@ -27,7 +27,7 @@ type##_retry(struct ip_set *set)					\
 	if (hashsize == map->hashsize)					\
 		hashsize++;						\
 									\
-	ip_set_printk("rehashing of set %s triggered: "			\
+	DP("rehashing of set %s triggered: "				\
 		      "hashsize grows from %lu to %lu",			\
 		      set->name,					\
 		      (long unsigned)map->hashsize,			\
