--- tc.func.c.orig      Tue Nov 22 21:54:05 1994
+++ tc.func.c   Tue Nov 22 21:37:43 1994
@@ -1877,7 +1877,8 @@
     else {
        char *ptr, *sptr;
        char *name = utmphost();
-       if (name != NULL) {
+       /* check agains " " to catch xterm bug... */
+       if (name != NULL && strcmp(name," ") != 0) {
            /* Look for host:display.screen */
            if ((sptr = strchr(name, ':')) != NULL)
                *sptr = '\0';
