--- freeradius-server-2.1.6/src/modules/rules.mak.orig	2009-06-23 13:09:52.000000000 +1000
+++ freeradius-server-2.1.6/src/modules/rules.mak	2009-06-23 13:09:54.000000000 +1000
@@ -40,8 +40,9 @@
 # along with a number of other useful definitions.
 #
 #######################################################################
-LT_OBJS		+= $(SRCS:.c=.lo)
-LT_OBJS		+= $(SRCS:.cpp=.lo)
+LT_TMP		:= $(SRCS:.c=.lo)
+LT_TMP		:= $(LT_TMP:.cpp=.lo)
+LT_OBJS		+= $(LT_TMP)
 CFLAGS		+= -I$(top_builddir)/src -I$(top_builddir)/libltdl
 
 #######################################################################
