--- freeradius-server-2.1.6/src/main/event.c.orig	2009-06-23 13:21:33.000000000 +1000
+++ freeradius-server-2.1.6/src/main/event.c	2009-06-23 13:21:35.000000000 +1000
@@ -1667,7 +1667,9 @@
 	 */
 	request->num_proxied_requests = 1;
 	request->num_proxied_responses = 0;
+#ifdef HAVE_PTHREAD_H
 	request->child_pid = NO_SUCH_CHILD_PID;
+#endif
 
 	update_event_timestamp(request->proxy, request->proxy_when.tv_sec);
 
