--- freeradius-server-2.1.6/src/modules/rlm_eap/Makefile.in.orig	2009-07-07 00:08:22.000000000 +1000
+++ freeradius-server-2.1.6/src/modules/rlm_eap/Makefile.in	2009-07-07 00:08:24.000000000 +1000
@@ -6,7 +6,7 @@
 SRCS        = rlm_eap.c eap.c mem.c
 HEADERS     = eap.h rlm_eap.h
 RLM_CFLAGS  = $(INCLTDL) -Ilibeap
-CLIENTLIBS  = libeap/$(LIBPREFIX)freeradius-eap.la
+CLIENTLIBS  = libeap/$(LIBPREFIX)freeradius-eap.la $(LIBRADIUS)
 RLM_LIBS    = libeap/$(LIBPREFIX)freeradius-eap.la $(LIBLTDL)
 RLM_INSTALL = install-subdirs
 RLM_SUBDIRS = libeap @eaptypes@
